src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1402,14 +1404,44 @@ private static class Key {
 
             public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
                 this.attrs = attrs;
-                this.file = file;
+                Object fk = attrs.fileKey();
+                if (fk != null) {
+                    // use file value, hashCode will use fileKey()
+                    this.file = file;
+                } else {
+                    this.file = getCanonicalFile(file);
+                }
                 this.utf8 = zc.isUTF8();
             }
 
+            @SuppressWarnings("removal")
+            private File getCanonicalFile(File file) {
+                if (System.getSecurityManager() != null) {
+                    PrivilegedAction<File> pa = () -> {
+                        try {
+                            return file.getCanonicalFile();
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
+                    }};
+                    return AccessController.doPrivileged(pa);
+                } else {
+                    try {
+                        return file.getCanonicalFile();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            }
+
             public int hashCode() {
                 long t = utf8 ? 0 : Long.MAX_VALUE;
                 t += attrs.lastModifiedTime().toMillis();
-                return ((int)(t ^ (t >>> 32))) + file.hashCode();
+                Object fk = attrs.fileKey();
+                if (fk != null) {
+                    return ((int)(t ^ (t >>> 32))) + fk.hashCode();
+                } else {
+                    return ((int)(t ^ (t >>> 32))) + file.hashCode();
+                }
```

Comment:
Consider this trivial simplification:
```suggestion
                return Long.hashCode(t)
                        + (fk != null ? fk.hashCode() : file.hashCode());
```